auditlog/log-enricher/src/main/java/io/debezium/demos/auditing/enricher/TopologyProducer.java

```diff
@@ -0,0 +1,278 @@
+package io.debezium.demos.auditing.enricher;
+
+import java.time.Duration;
+import java.util.Optional;
+
+import javax.enterprise.context.ApplicationScoped;
+import javax.enterprise.inject.Produces;
+import javax.json.Json;
+import javax.json.JsonObject;
+
+import org.apache.kafka.common.serialization.Serdes;
+import org.apache.kafka.streams.KeyValue;
+import org.apache.kafka.streams.StreamsBuilder;
+import org.apache.kafka.streams.Topology;
+import org.apache.kafka.streams.kstream.Materialized;
+import org.apache.kafka.streams.kstream.Transformer;
+import org.apache.kafka.streams.processor.ProcessorContext;
+import org.apache.kafka.streams.processor.PunctuationType;
+import org.apache.kafka.streams.state.KeyValueStore;
+import org.apache.kafka.streams.state.StoreBuilder;
+import org.apache.kafka.streams.state.Stores;
+import org.eclipse.microprofile.config.inject.ConfigProperty;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+@ApplicationScoped
+public class TopologyProducer {
+
+    private static final Logger LOG = LoggerFactory.getLogger(TopologyProducer.class);
+
+    private static final Long BUFFER_OFFSETS_KEY = -1L;
+    private static final String STREAM_BUFFER_NAME = "stream-buffer-state-store";
+    private static final String STORE_NAME = "transaction-meta-data";
+
+
+    @ConfigProperty(name = "audit.context.data.topic")
+    String txContextDataTopic;
+
+    @ConfigProperty(name = "audit.vegetables.topic")
+    String vegetablesTopic;
+
+    @ConfigProperty(name = "audit.vegetables.enriched.topic")
+    String vegetablesEnrichedTopic;
+
+    @Produces
+    public Topology buildTopology() {
+        StreamsBuilder builder = new StreamsBuilder();
+
+        StoreBuilder<KeyValueStore<Long, JsonObject>> streamBufferStateStore =
+                Stores
+                    .keyValueStoreBuilder(
+                        Stores.persistentKeyValueStore(STREAM_BUFFER_NAME),
+                        new Serdes.LongSerde(),
+                        new JsonObjectSerde()
+                    )
+                    .withCachingDisabled();
+            builder.addStateStore(streamBufferStateStore);
+
+        builder.globalTable(txContextDataTopic, Materialized.as(STORE_NAME));
+
+        builder.<JsonObject, JsonObject>stream(vegetablesTopic)
+                // filter out any tombstones; for an audit log topic, time-based retention
+                // seems more reasonable than compaction
+                .filter((id, changeEvent) -> changeEvent != null)
+                // exclude snapshot events
+                .filter((id, changeEvent) -> !changeEvent.getString("op").equals("r"))
+                // enrich change events with transaction metadata via the statestore of the TX topic
+                .transform(() -> new ChangeEventEnricher(), STREAM_BUFFER_NAME)
```

Comment:
@gunnarmorling Just one technical note, you are using Transformer albeit you are not modifying the key. Using transforme can lead to repartitioning as it is expected that the key was modified. Processor does not have this problem.